crates/proto/src/serialize/binary/encoder.rs

```diff
@@ -89,8 +89,8 @@ mod private {
 pub struct BinEncoder<'a> {
     offset: usize,
     buffer: private::MaximalBuf<'a>,
-    /// start and end of label pointers, smallvec here?
-    name_pointers: Vec<(usize, usize)>,
+    /// start of label pointers with their labels in fully decompressed form for easy comparison, smallvec here?
+    name_pointers: Vec<(usize, Vec<u8>)>,
```

Comment:
This might be complex, but could we potentially change this to a slice? I -think- it could be `&'a [u8]`, without too much issue?

Reply:
Sounds good. Let me give it a try. Thanks for the suggestion.

Reply:
I did another look into this. The reason why it has to be a vector is because the content does not live in the buffer, so this |name_pointers| is the owner of the data.

For example:
Lets say the order of names we see is as follows:
1. "ra.rb.rc"
2. "ra.rc"
3. "ra.rc"

When we see a name "ra.rb.rc", we put something like "2ra2rb2rc " in the buffer.

Then when we see "ra.rc", we append "2ra2rc" and then we see that we have "rc" in the buffer, so we go back and wipe out "2rc" and instead insert a two bytes pointer there. Now, the buffer looks something like "2ra2rb2rc 2ra[two bytes of pointer]", so we "lose" the slice for "ra.rc" when we go to the third name "ra.rc".

Does that make sense?

Reply:
The signature I was thinking of is `name_pointers: Vec<(usize, &'a [u8])>`

This would be pushed in via `self.name_pointers.push((start, self.slice_of(start, end)))`

I just want to make sure we're talking about the same thing?

Reply:
Ok, I just pulled your branch to see what's going on, I can see it's not simple. The ownership does seem to be an issue.

Reply:
I was worried this might represent a performance degradation, but it doesn't appear to be noticeable, and may in fact be better.

I used the benchmarks in `/bin/benches`: `cargo +nightly bench trust_dns_udp_bench -- --nocapture` and `cargo +nightly bench trust_dns_tcp_bench -- --nocapture`

Reply:
I had this concern as well; however, using pointer based solution requires some form of de-referencing to appropriately [compare the labels](https://github.com/bluejekyll/trust-dns/blob/main/crates/proto/src/serialize/binary/encoder.rs#L254). I am working on doing some benchmark on my end as well to see if we have any regression here. Thanks for helping to do the benchmarks and get a promising result! :)